src/Xml/XmlDocument.php

```diff
@@ -48,79 +44,118 @@
 use const LIBXML_ERR_WARNING;
 
 /**
- * Class for validating OPUS import xml.
+ * A class for loading and validating XML
+ * (replaces/merges two previous separate validation classes).
  */
-class XmlValidation
+class XmlDocument
 {
+    /** @var DOMDocument|null */
+    private $xml;
+
     /** @var array|null */
     private $errors;
 
     /**
-     * Validates import XML based on schema.
-     *
-     * Can validate DOMDocument objects or XML provided as string.
+     * Constructor of the class
      *
-     * @param DOMDocument|string $xml
-     * @return bool
+     * @param DOMDocument|null $doc
      */
-    public function validate($xml)
+    public function __construct($doc = null)
     {
-        $this->errors = null;
-
-        // TODO check for null|empty
-
-        if (! $xml instanceof DOMDocument) {
-            $xml = $this->getDocument($xml);
+        if ($doc) {
+            $this->xml = $doc;
         }
+    }
 
+    /**
+     * Loads an XML string
+     *
+     * @param  string $xml XML string
+     * @return DOMDocument
+     */
+    public function loadXML($xml)
+    {
+        // Enable user error handling
         libxml_clear_errors();
-        libxml_use_internal_errors(true);
-
-        $valid = $xml->schemaValidate(__DIR__ . DIRECTORY_SEPARATOR . 'Xml' . DIRECTORY_SEPARATOR . 'opus-import.xsd');
+        $useInternalErrors = libxml_use_internal_errors(true);
 
+        $doc          = new DOMDocument();
+        $success      = $doc->loadXML($xml);
         $this->errors = libxml_get_errors();
+
+        // Disable user error handling
+        libxml_use_internal_errors($useInternalErrors);
         libxml_clear_errors();
-        libxml_use_internal_errors(false);
 
-        return $valid;
-    }
+        if (! $success) {
+            throw new MetadataImportInvalidXmlException($this->getErrorMessage());
+        }
 
-    /**
-     * @return array|null
-     */
-    public function getErrors()
-    {
-        return $this->errors;
+        $this->xml = $doc;
+        return $doc;
     }
 
     /**
-     * @param string $xml
+     * Loads XML from a file
+     *
+     * @param  string $xmlFilePath Path to the xml file
      * @return DOMDocument
      */
-    private function getDocument($xml)
+    public function load($xmlFilePath)
```

Comment:
Es ging darum, dass `XmlDocument` die beiden `DomDocument` Funktionen `load` und `loadXML` spiegelt, anstatt ein Parameter für die Unterscheidung zwischen Strings und Dateien zu verwenden. Zwei Identische Funktionen machen aber keinen Sinn und PHPCPD würde vermutlich meckern. Wie kann man das fixen? 

Vermutlich kommt man nicht darum herum die eigentliche Lade-Operation in eine interne (protected) Funktion auszulagern, die dann wieder einen Parameter verwendet. Oder man könnte in der Funktionen, die Dateien lädt, die Datei zuerst in einen String laden. Das ist aber unter Umständen weniger effizient, abhängig davon, wie die `load` für `DOMDocument` implementiert wurde. Im Augenblick wäre mir das egal.

Reply:
Ich habe jetzt eine interne Funktion mit Parameter eingebaut.